src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,203 @@
+-/
+def rev_at_fun (N i : ℕ) : ℕ := ite (i ≤ N) (N-i) i
+
+lemma rev_at_fun_invol {N i : ℕ} : rev_at_fun N (rev_at_fun N i) = i :=
+begin
+  unfold rev_at_fun,
+  split_ifs with h j,
+  { exact nat.sub_sub_self h, },
+  { exfalso,
+    apply j,
+    exact nat.sub_le N i, },
+  { refl, },
+end
+
+lemma rev_at_fun_inj {N : ℕ} : function.injective (rev_at_fun N) :=
+begin
+  intros a b hab,
+  rw [← @rev_at_fun_invol N a, hab, rev_at_fun_invol],
+end
+
+/-- If `i ≤ N`, then `rev_at N i` returns `N - i`, otherwise it returns `i`.
+Essentially, this embedding is only used for `i ≤ N`.
+The advantage of `rev_at N i` over `N - i` is that `rev_at` is an involution.
+-/
+def rev_at (N : ℕ) : function.embedding ℕ ℕ :=
+{ to_fun := λ i , (ite (i ≤ N) (N-i) i),
+  inj' := rev_at_fun_inj }
+
+/-- We prefer to use the bundled `rev_at` over unbundled `rev_at_fun`. -/
+@[simp] lemma rev_at_fun_eq (N i : ℕ) : rev_at_fun N i = rev_at N i := rfl
+
+@[simp] lemma rev_at_invol {N i : ℕ} : (rev_at N) (rev_at N i) = i :=
+rev_at_fun_invol
+
+@[simp] lemma rev_at_le {N i : ℕ} (H : i ≤ N) : rev_at N i = N - i :=
+if_pos H
+
+lemma rev_at_add {N O n o : ℕ} (hn : n ≤ N) (ho : o ≤ O) :
+  rev_at (N + O) (n + o) = rev_at N n + rev_at O o :=
+begin
+  rcases nat.le.dest hn with ⟨n', rfl⟩,
+  rcases nat.le.dest ho with ⟨o', rfl⟩,
+  repeat { rw rev_at_le (le_add_right rfl.le) },
+  rw [add_assoc, add_left_comm n' o, ← add_assoc, rev_at_le (le_add_right rfl.le)],
+  repeat {rw nat.add_sub_cancel_left},
+end
+
+/-- `reflect N f` is the polynomial such that `(reflect N f).coeff i = f.coeff (rev_at N i)`.
+In other words, the terms with exponent `[0, ..., N]` now have exponent `[N, ..., 0]`.
+
+In practice, `reflect` is only used when `N` is at least as large as the degree of `f`.
+
+Eventually, it will be used with `N` exactly equal to the degree of `f`.  -/
+noncomputable def reflect (N : ℕ) (f : polynomial R) : polynomial R :=
+finsupp.emb_domain (rev_at N) f
+
+lemma reflect_support (N : ℕ) (f : polynomial R) :
+  (reflect N f).support = image (rev_at N) f.support :=
+begin
+  ext1,
+  rw [reflect, mem_image, support_emb_domain, mem_map],
+end
+
+@[simp] lemma coeff_reflect (N : ℕ) (f : polynomial R) (i : ℕ) :
+  coeff (reflect N f) i = f.coeff (rev_at N i) :=
+calc finsupp.emb_domain (rev_at N) f i
+    = finsupp.emb_domain (rev_at N) f (rev_at N (rev_at N i)) : by rw rev_at_invol
+... = f.coeff (rev_at N i) : finsupp.emb_domain_apply _ _ _
+
+@[simp] lemma reflect_zero {N : ℕ} : reflect N (0 : polynomial R) = 0 := rfl
+
+@[simp] lemma reflect_eq_zero_iff {N : ℕ} {f : polynomial R} :
+  reflect N (f : polynomial R) = 0 ↔ f = 0 :=
+begin
+  split,
+  { intros a,
+    injection a with f0 f1,
+    rwa [map_eq_empty, support_eq_empty] at f0, },
+  { rintro rfl,
+    refl, },
+end
+
+@[simp] lemma reflect_add (f g : polynomial R) (N : ℕ) :
+  reflect N (f + g) = reflect N f + reflect N g :=
+by { ext1, rw [coeff_add, coeff_reflect, coeff_reflect, coeff_reflect, coeff_add], }
```

Comment:
Do proofs like this not work by `by { ext, simp }`? If `simp` handles it (which I'd hope it should, if you have enough simp lemmas), it just seems clutter to spell out the steps.

Reply:
Wouldn't we want to `squeeze_simp` the result anyway? I've seen quite a bunch of slowish proofs recently that could be sped up using `squeeze_simp`.
This brings me back to my point of wishing for two layers of proof presentations: the once that show how Lean can *find* a proof itself, and those where Lean only really has to do verification (even if it uses tactics).

Reply:
You are right that ```{ ext, simp }``` completes the proof, but I strive as much as possible to avoid using ```simp```.  If you believe that this is a mistake, I can leave ```simp```s in proofs, but I was explicitly replacing ```simp```s by step-by-step proofs.

Reply:
A quick, unscientific experiment using `set_option profiler true` shows the following times on my machine:

 * `by { ext1, rw }`: 9 ms tactic execution
 * `by { ext, simp only [coeff_add, coeff_reflect] }`: 22 ms tactic execution
 * `by { ext, simp }`: 231 ms tactic execution

231 milliseconds is already large enough to notice, so I would vote for keeping them as `rw` or, as a compromise, using `simp only`.

Reply:
I am very happy to have learnt about ```set_option profiler true``` and that it supports my coding idiosyncrasies...

Reply:
Still, I am happy to comply with alternative conventions, should this be preferred!

Reply:
If a `rw` mentions some lemmas multiple times, I usually opt for the `simp only` for brevity.